### ISSUE TITLE
ArgoCD deleted all applications during the GitHub service degradation

### ISSUE THREAD

**Issue by cazorla19** [argoproj/argo-cd#3473]
**Describe the bug**

At one moment, I encountered the situation described by every WebOps book but ignored by most engineers – complete wipeout of all apps on staging/production deployed via ArgoCD.

After systems recovery and further debugging, I see that the event that caused the issue was the GitHub downtime that happened at the same time.

![image](https://user-images.githubusercontent.com/14975946/80118509-caa3cb80-8588-11ea-8d5e-9e195b9a7a4b.png)

At the moment, I have the following picture in my mind:

1. ArgoCD polled gitops repository hosted in GitHub every 3 minutes, as usual
2. At the last polling, ArgoCD fetched the latest commit that one of our applications (let's call it `backend`) has been updated and needs to be deployed (staging) or put the deployment on hold (production)
3. ArgoCD started to look for all repositories to fetch the Helm chart from each one of them, but it couldn’t – because GitHub was down at that moment (https://www.githubstatus.com/)
4. No Helm charts – no applications, ArgoCD decides to delete all apps as it can’t fetch their Helm charts

We run ArgoCD on 3 environments via the same gitops repository, but the third one hadn't wiped all apps out, as we didn't register `backend` there that caused the environment diff.

Interesting fact, at the step 2 `argocd-repo-server` was able to fetch the latest commit from `gitops` repo, which means GitHub was available – and when it started to apply the changes, it wasn’t.

**To Reproduce**

I'm not sure if it's possible, because not only GitHub downtime needed, but the ability to fetch the commit before downtime needs to be in the reproduction too.

**Expected behavior**

ArgoCD should have any(!) fallback mechanism for such cases, but it shouldn't wipe out all apps when the GitHub was unreachable.

**Screenshots**

A history log that triggered the update

![image](https://user-images.githubusercontent.com/14975946/80122362-b9a98900-858d-11ea-98e2-c06eb0861419.png)

ArgoCD Logs from DataDog. The same sequence [up to deletion] happened with all apps. Unfortunately, I have only a screenshot of logs as the ArgoCD has been wiped out himself as the dependency of one helm chart [yes, "chicken and egg" problem].

![image](https://user-images.githubusercontent.com/14975946/80121744-fb85ff80-858c-11ea-8a04-033b8cc3b3f5.png)

**Version**

```shell
bash-5.0# argocd version
argocd: v1.4.3+2483411
  BuildDate: 2020-04-15T16:48:21Z
  GitCommit: 24834112d66f453de6ae53901d5dad0ad71ec902
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
argocd-server: v1.4.2+48cced9
  BuildDate: 2020-01-24T01:07:03Z
  GitCommit: 48cced9d925b5bc94f6aa9fa4a8a19b2a59e128a
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: v2.15.2
  Kubectl Version: v1.14.0
```

The app pack has been deployed via `app-of-apps` entity.

**Author note**

I don't really want to shout `It's critical!!!`, but assuming that all apps being synced by stable version were wiped out, I'd say it is. Let me know if you need any other input from my side

**Comment by alexmt** [comment]
I think it is critical. The same issue was reported before https://github.com/argoproj/argo-cd/issues/2573 but I never could reproduce it.

I want to make sure I got it right:

* You using App-of-apps pattern and Argo CD deleted parent app resources (applications)?
* If App-of-apps app is plain YAML directory?
* Can you please attach argocd-repo-server logs ?

**Comment by alexmt** [comment]
@cazorla19 I would really appreciate if you can join our [slack](https://argoproj.github.io/community/join-slack) and help to reproduce this issue

**Comment by cazorla19** [comment]
Hi, @alexmt, thanks for your quick response.

Answering your questions:

* Correct
* It's the Helm chart directory, e.g. the app-of-apps manifest is being templated by Helm
* Unfortunately, `argocd-repo-server` was wiped out as well as it's been deployed as the part of one ArgoCD app so I couldn't get logs

Thanks for inviting me to Slack, I've signed up as `@Misha Chinkov`. Will try to help you next week as I still need to recover the gitops state and post-mortem this incident

**Comment by cazorla19** [comment]
Hi, @alexmt, thanks for the quick response.

After the detailed investigation, I found out that it was rather the problem of tooling around ArgoCD than the ArgoCD itself.

We're using git-bot in CircleCI to update application revisions after they're built and tested in `master` branch. The script around the git-bot tried to fetch `yq` utility to update the YAML file with tags/revisions, but it failed. Instead of failing the script, it wiped out the whole YAML file and pushed the empty one to `master` of gitops repository.

And even there I could prevent the wipeout by disabling `auto-prune` in production, but I didn't. So the auto-prune feature shot my leg, so to say.

In other words, the ArgoCD didn't have any bug on that. I'm very sorry for pushing this issue without complete investigation. Can I close this ticket then?

**Comment by alexmt** [comment]
Thank you for update @cazorla19 ! No problem at all. Let's close this ticket. Opening another one to add some protection from automation/human error https://github.com/argoproj/argo-cd/issues/3484
